worker/externalcontrollerupdater/externalcontrollerupdater.go

```diff
@@ -0,0 +1,266 @@
+func New(
+	externalControllers ExternalControllerUpdaterClient,
+	newExternalControllerWatcherClient NewExternalControllerWatcherClientFunc,
+	clock clock.Clock,
+) (worker.Worker, error) {
+	w := updaterWorker{
+		watchExternalControllers:           externalControllers.WatchExternalControllers,
+		externalControllerInfo:             externalControllers.ExternalControllerInfo,
+		setExternalControllerInfo:          externalControllers.SetExternalControllerInfo,
+		newExternalControllerWatcherClient: newExternalControllerWatcherClient,
+		runner: worker.NewRunner(worker.RunnerParams{
+			// One of the controller watchers fails should not
+			// prevent the others from running.
+			IsFatal: func(error) bool { return false },
+
+			// If the API connection fails, try again in 1 minute.
+			RestartDelay: time.Minute,
+			Clock:        clock,
+		}),
+	}
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+		Init: []worker.Worker{w.runner},
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+	return &w, nil
+}
+
+type updaterWorker struct {
+	catacomb catacomb.Catacomb
+	runner   *worker.Runner
+
+	watchExternalControllers           func() (watcher.StringsWatcher, error)
+	externalControllerInfo             func(controllerUUID string) (*crossmodel.ControllerInfo, error)
+	setExternalControllerInfo          func(crossmodel.ControllerInfo) error
+	newExternalControllerWatcherClient NewExternalControllerWatcherClientFunc
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *updaterWorker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *updaterWorker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *updaterWorker) loop() error {
+	watcher, err := w.watchExternalControllers()
+	if err != nil {
+		return errors.Annotate(err, "watching external controllers")
+	}
+	w.catacomb.Add(watcher)
+
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+
+		case ids, ok := <-watcher.Changes():
+			if !ok {
+				return w.catacomb.ErrDying()
+			}
+
+			if len(ids) == 0 {
+				continue
+			}
+
+			logger.Debugf("external controllers changed: %q", ids)
+			tags := make([]names.ControllerTag, len(ids))
+			for i, id := range ids {
+				if !names.IsValidController(id) {
+					return errors.Errorf("%q is not a valid controller tag", id)
+				}
+				tags[i] = names.NewControllerTag(id)
+			}
+
+			watchers := make(set.Tags)
+			for _, tag := range tags {
+				// We're informed when an external controller
+				// is added or removed, so treat as a toggle.
+				if watchers.Contains(tag) {
+					logger.Infof("stopping watcher for external controller %q", tag.Id())
+					w.runner.StopWorker(tag.Id())
+					watchers.Remove(tag)
+					continue
+				}
+				logger.Infof("starting watcher for external controller %q", tag.Id())
+				watchers.Add(tag)
+				w.runner.StartWorker(tag.Id(), func() (worker.Worker, error) {
```

Comment:
The error from StartWorker should not be ignored.

Reply:
fixed